File: server/vesnai/ai/tagging.py
"""On-demand tag and type suggestions for new notes."""

from __future__ import annotations

import json
import re
from typing import TYPE_CHECKING

from vesnai.providers.base import AIProvider

if TYPE_CHECKING:
    from vesnai.ai.selftune import TagClassifier

_WORD = re.compile(r"\w+", re.UNICODE)


def heuristic_tags(title: str, body: str) -> tuple[str, list[str]]:
    text = f"{title} {body}".lower()
    keywords = {
        "idea": ["idea", "maybe", "what if", "could", "startup"],
        "travel": ["trip", "travel", "flight", "visit", "vacation"],
        "todo": ["todo", "buy", "remember", "task", "call", "email"],
        "photo": ["photo", "picture", "image", "selfie"],
    }
    tags: list[str] = []
    for tag, words in keywords.items():
        if any(w in text for w in words):
            tags.append(tag)
    if not tags:
        tags = ["misc"]
    note_type = "Note"
    if "idea" in tags:
        note_type = "Idea"
    elif "photo" in tags:
        note_type = "Photo"
    elif "todo" in tags:
        note_type = "Task"
    return note_type, tags


def suggest_tags(
    ai: AIProvider,
    *,
    title: str,
    body: str,
    known_tags: list[str] | None = None,
    tag_classifier: TagClassifier | None = None,
) -> dict[str, object]:
    """Return ``{type, tags}`` using classifier, reasoning model, or heuristics."""
    text = f"{title} {body}".strip()
    if tag_classifier is not None and tag_classifier.is_trained:
        predicted = tag_classifier.predict(text, top_k=4)
        if predicted:
            note_type, _ = heuristic_tags(title, body)
            merged = list(dict.fromkeys([*predicted, *(known_tags or [])[:2]]))[:6]
            return {"type": note_type, "tags": merged}
    hint = ""
    if known_tags:
        hint = f"\nPrefer tags from this vocabulary when relevant: {', '.join(known_tags[:30])}."
    prompt = (
        "Suggest a note type and 3-6 lowercase tags for this personal note. "
        "Reply with JSON only: {\"type\": \"...\", \"tags\": [\"...\"]}."
        f"{hint}\n\nTitle: {title}\nBody: {body}"
    )
    try:
        raw = ai.complete(prompt, temperature=0.2, think=True).strip()
        if raw.startswith("```"):
            raw = raw.split("\n", 1)[-1].rsplit("```", 1)[0].strip()
        data = json.loads(raw)
        tags = [str(t).lower().strip() for t in data.get("tags", []) if str(t).strip()]
        note_type = str(data.get("type") or "Note").strip() or "Note"
        if not tags:
            note_type, tags = heuristic_tags(title, body)
        return {"type": note_type, "tags": tags[:8]}
    except Exception:
        note_type, tags = heuristic_tags(title, body)
        return {"type": note_type, "tags": tags}

File: app/lib/data/tagging.dart
/// On-device tag suggestion.
///
/// Behind an interface so it is mockable and so the heuristic baseline can later
/// be swapped for an on-device ML model without touching callers. Suggestions
/// are user-editable; accept/reject signals feed the server's self-tuning loop.
abstract class Tagger {
  /// Suggest a `(type, tags)` pair for the given note content.
  TagSuggestion suggest(String title, String body);
}

class TagSuggestion {
  final String type;
  final List<String> tags;
  const TagSuggestion(this.type, this.tags);
}

class HeuristicTagger implements Tagger {
  const HeuristicTagger();

  static final Map<String, List<String>> _keywords = {
    'idea': ['idea', 'maybe', 'what if', 'could', 'startup', 'invent'],
    'travel': ['trip', 'travel', 'flight', 'visit', 'vacation'],
    'todo': ['todo', 'buy', 'remember', 'task', 'call', 'email'],
    'photo': ['photo', 'picture', 'image', 'selfie'],
    'misc': [],
  };

  @override
  TagSuggestion suggest(String title, String body) {
    final text = '$title $body'.toLowerCase();
    final tags = <String>[];
    for (final entry in _keywords.entries) {
      if (entry.value.any(text.contains)) tags.add(entry.key);
    }
    if (tags.isEmpty) tags.add('misc');

    String type = 'Note';
    if (tags.contains('idea')) {
      type = 'Idea';
    } else if (tags.contains('photo')) {
      type = 'Photo';
    } else if (tags.contains('todo')) {
      type = 'Task';
    }
    return TagSuggestion(type, tags);
  }
}
